File: DAYS/Day04-ProductExceptSelf/ProductExceptSelf.py
# Array Product Except Self
# given an integer array nums, return an array answer such that answer[i] 
# is equal to the product of all the elements of nums except nums[i].

# Approach: Prefix and Suffix Product
class soultion:

    def productExceptSelf(self,nums):
        n=len(nums)
        res=[1]*n
        prefix=1  # prefix is the product of all the elements to the left of the current index
        for i in range(n):
            res[i]*=prefix
            prefix *=nums[i]
        suffix=1  # suffix is the product of all the elements to the right of the current index
        for i in range(n-1,-1,-1):
            res[i]*=suffix
            suffix*=nums[i]
        return res

# test cases
print(soultion().productExceptSelf([1,2,3,4]))
# output: [24,12,8,6]
print(soultion().productExceptSelf([-1,1,0,-3,3]))
# output: [0,0,9,0,0]

# Complexity Analysis:
# Brute Force Approach: Time Complexity: O(n^2) , Space Complexity: O(1)
# Prefix and Suffix Product: Time Complexity: O(n), Space Complexity: O(n)

File: DAYS/Day04-ProductExceptSelf/ProductExceptSelf.js
// Array problem - Product of Array Except Self
// Given an integer array nums, return an array answer such that answer[i]
//  is equal to the product of all the elements of nums except nums[i].

function productExceptSelf(nums) {
    const n=nums.length;
    const res=new Array(n).fill(1)
    let prefix=1 // prefix[i] stores the product of numbers from 0 to i-1
    for (let i=0;i<n;i++){
        res[i] *= prefix // res[i] is the product of numbers before nums[i]
        prefix *= nums[i]
    }
    let suffix=1 // suffix[i] stores the product of numbers from i+1 to n-1
    for(let i=n-1;i>=0;i--){
        res[i] *=suffix // res[i] is the product of all numbers except nums[i]
        suffix *=nums[i]
    }
    return res
}
// Test cases
console.log(productExceptSelf([1,2,3,4])) // [24,12,8,6]
console.log(productExceptSelf([-1,1,0,-3,3])) // [-0,0,9,-0,0]

// complexity analysis :
// time complexity : O(n)
// space complexity : O(n)